### PR TITLE
simple-linked-list: add exercise

### LINKED ISSUE #931: simple-linked-list: implement exercise
[Specification](https://github.com/exercism/problem-specifications/tree/master/exercises/simple-linked-list)

### PR DESCRIPTION
**Pull request by ngochai94**
Resolves #931

### CODE REVIEW

File: exercises/simple-linked-list/example.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
+package linkedlist
+
+import "fmt"
+
+type Element struct {
```
Comment (ferhatelmas) [inline exercises/simple-linked-list/example.go]: do we need to export ?

File: exercises/simple-linked-list/example.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+	next *Element
+}
+
+type List struct {
+	head *Element
+	size int
+}
+
+func New(array []int) *List {
+	res := List{nil, 0}
```
Comment (ferhatelmas) [inline exercises/simple-linked-list/example.go]: defaults are fine so `var res List`

File: exercises/simple-linked-list/example.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+			node = node.next
+		}
+		node.next = &newNode
+	}
+	list.size++
+}
+
+func (list *List) Pop() error {
+	if list.size == 0 {
+		return fmt.Errorf("Cannot call pop on empty list")
```
Comment (ferhatelmas) [inline exercises/simple-linked-list/example.go]: no formatting directive so `errors.New` directly

File: exercises/simple-linked-list/example.go [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+		node.next = nil
+	}
+	list.size--
+	return nil
+}
+
+func (list *List) Array() []int {
+	res := []int{}
+	node := list.head
+	for node != nil {
```
Comment (ferhatelmas) [inline exercises/simple-linked-list/example.go]: `for node := list.head; node != nil; node = node.next {`

File: exercises/simple-linked-list/example.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+		res = append(res, node.data)
+		node = node.next
+	}
+	return res
+}
+
+func (list *List) Reverse() *List {
+	// not the best algo but it works XD
```
Comment (ferhatelmas) [inline exercises/simple-linked-list/example.go]: let's remove it since we know it's an example.

File: exercises/simple-linked-list/example.go [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+		node := list.head
+		for node.next != nil {
+			node = node.next
+		}
+		node.next = &newNode
+	}
+	list.size++
+}
+
+func (list *List) Pop() error {
```
Comment (ferhatelmas) [inline exercises/simple-linked-list/example.go]: might be useful to return popped value

File: exercises/simple-linked-list/linked_list_test.go [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+
+import (
+	"reflect"
+	"testing"
+)
+
+var array1To10 = []int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}
+
+func TestEmptyListHasSizeZero(t *testing.T) {
+	list := New([]int{})
```
Comment (ferhatelmas) [inline exercises/simple-linked-list/linked_list_test.go]: how about `New(nil)` too ?

File: exercises/simple-linked-list/linked_list_test.go [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+	size := list.Size()
+	if size != 0 {
```
Comment (ferhatelmas) [inline exercises/simple-linked-list/linked_list_test.go]: `if size := list.Size(); size != 0 {`

File: exercises/simple-linked-list/linked_list_test.go [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+		t.Errorf("Reverse empty list: %v, want empty list", reversed)
+	}
+}
+
+func TestReverseNonEmptyList(t *testing.T) {
+	list := New([]int{1, 2, 3})
+	actual := list.Reverse().Array()
+	expected := []int{3, 2, 1}
+	if !reflect.DeepEqual(actual, expected) {
+		t.Errorf("Reverse non empty list: %v, want %v", actual, expected)
```
Comment (ferhatelmas) [inline exercises/simple-linked-list/linked_list_test.go]: Since there is one case, it doesn't matter here but we prefer `Fatalf`. Could you please use it ?

File: exercises/simple-linked-list/example.go (hunk 1 above)
Comment (ngochai94) [reply exercises/simple-linked-list/example.go]: right, good point 👍

File: exercises/simple-linked-list/example.go (hunk 6 above)
Comment (ngochai94) [reply exercises/simple-linked-list/example.go]: sure, I will change this and update the tests

File: exercises/simple-linked-list/example.go (hunk 3 above)
Comment (ngochai94) [reply exercises/simple-linked-list/example.go]: right, but I'm wondering why `errors.New` is preferred over `fmt.Errorf`? For me, I always use `fmt.Errorf` in case of later I might need to format the error message and save 1 import

File: exercises/simple-linked-list/example.go (hunk 4 above)
Comment (ngochai94) [reply exercises/simple-linked-list/example.go]: 👍

File: exercises/simple-linked-list/linked_list_test.go (hunk 7 above)
Comment (ngochai94) [reply exercises/simple-linked-list/linked_list_test.go]: should we expect the solution to handle a nil input? In that case, should I add tests for all empty-list tests but with nil input?

File: exercises/simple-linked-list/linked_list_test.go (hunk 8 above)
Comment (ngochai94) [reply exercises/simple-linked-list/linked_list_test.go]: 👍

File: exercises/simple-linked-list/linked_list_test.go (hunk 7 above)
Comment (ferhatelmas) [reply exercises/simple-linked-list/linked_list_test.go]: I would expect it to behave similar since we didn't supply any element from the point of linked list.

**Comment by ngochai94** [comment]
Thanks @ferhatelmas for reviewing! I have made changes accordingly. Also, I have a question which might not be related to this PR, but I sometimes saw CI fails temporarily such as with `git clone` or fetching `configlet`. Is there a way that we could trigger the CI rebuild?

**Comment by ferhatelmas** [comment]
@ngochai94 push will trigger it. There is a restart button but I'm not sure if you can see it since you don't have write access to repo.

**Comment by ngochai94** [comment]
Thanks @ferhatelmas . Yes, I don't have access to the rebuild button. I'm wondering if we could setup a webhook to trigger the CI rebuild upon a fixed PR comment, such as `Retest this please`
